quickwit-storage/src/split.rs

```diff
@@ -0,0 +1,469 @@
+                    )
+                })?);
+                Ok((file_name, range.start..range.end))
+            })
+            .collect::<Result<HashMap<_, _>, io::Error>>()?;
+
+        let metadata_json = serde_json::to_string(&BundleStorageFileOffsets {
+            files: metadata_with_fixed_paths,
+        })?;
+
+        footer_bytes.extend(metadata_json.as_bytes());
+        let metadata_json_len = metadata_json.len() as u64;
+        footer_bytes.extend(&metadata_json_len.to_le_bytes());
+        footer_bytes.extend(hotcache);
+        footer_bytes.extend(&hotcache.len().to_le_bytes());
+
+        let mut payloads: Vec<Box<dyn PutPayload>> = Vec::new();
+
+        let mut sorted_files = self.metadata.files.iter().collect::<Vec<_>>();
+        sorted_files.sort_by_key(|(_file, range)| range.start);
+
+        for (path, byte_range) in sorted_files {
+            let file_payload = FilePayload {
+                path: path.to_owned(),
+                len: byte_range.end - byte_range.start,
+            };
+            payloads.push(Box::new(file_payload));
+        }
+
+        payloads.push(Box::new(footer_bytes.to_vec()));
+
+        Ok(SplitPayload {
+            payloads,
+            footer_range: self.current_offset as u64
+                ..self.current_offset as u64 + footer_bytes.len() as u64,
+        })
+    }
+}
+
+fn get_payload_with_range(
+    payloads: &[Box<dyn PutPayload>],
+) -> Vec<(Box<dyn PutPayload>, Range<usize>)> {
+    let mut current = 0;
+    payloads
+        .iter()
+        .map(|payload| {
+            let start = current;
+            current += payload.len();
+            (payload.clone(), start as usize..current as usize)
+        })
+        .collect()
+}
+
+fn do_ranges_overlap(range1: &Range<usize>, range2: &Range<usize>) -> bool {
+    !get_ranges_overlap(range1, range2).is_empty()
+}
+
+fn get_ranges_overlap(range1: &Range<usize>, range2: &Range<usize>) -> Range<usize> {
+    range1.start.max(range2.start)..range1.end.min(range2.end)
+}
+
+// Returns payloads and their relative ranges for an absolute range.
+fn chunk_payload_ranges(
+    payloads: &[Box<dyn PutPayload>],
+    range: Range<usize>,
+) -> Vec<(Box<dyn PutPayload>, Range<usize>)> {
+    let payload_with_range = get_payload_with_range(payloads);
+    let mut ranges = vec![];
+    let mut current_range = range;
+
+    while let Some((payload, payload_range)) = payload_with_range
+        .iter()
+        .find(|block| do_ranges_overlap(&block.1, &current_range))
+    {
+        let payload_range_absolute_end = current_range.end.min(payload_range.end);
+        let relative_range_in_payload = current_range.start - payload_range.start
+            ..payload_range_absolute_end - payload_range.start;
+
+        ranges.push((payload.clone(), relative_range_in_payload));
+
+        current_range = payload_range.end..current_range.end;
+    }
+
+    ranges
+}
```

Comment:
I struggled a little bit to review this function, I feel that this is a little bit clearer to compute the absolute ranges in it directly so that we see exactly what we are doing in this function. What do you think?
You can update the docs too.

```rust
/// Returns payloads and their relative ranges that overlap with the input `range`.
fn chunk_payload_ranges(
    payloads: &[Box<dyn PutPayload>],
    range: Range<usize>,
) -> Vec<(Box<dyn PutPayload>, Range<usize>)> {
    let mut ranges = vec![];
    let mut current_payload_absolute_start = 0usize;
    for payload in payloads {
        let current_payload_absolute_end = current_payload_absolute_start + payload.len() as usize;
        let absolute_range_overlap = get_ranges_overlap(&(current_payload_absolute_start..current_payload_absolute_end), &range);
        if !absolute_range_overlap.is_empty() {
            // Push the range relative to this payload as we will read from it.
            ranges.push((payload.clone(), (absolute_range_overlap.start - current_payload_absolute_start)..(absolute_range_overlap.end - current_payload_absolute_start)));
        }
        current_payload_absolute_start = current_payload_absolute_end;
    }
    ranges
}
```

Reply:
(ps I tested this function)

Reply:
yes this is easier to read